File: frontend/src/types/customer.ts
export type CustomerStatus = 'active' | 'inactive' | 'watch'

export type PurchaseRow = {
  id: string
  date: string
  amount: number
  reference: string
  channel: 'counter' | 'delivery' | 'phone'
}

export type DebtRow = {
  id: string
  date: string
  type: 'charge' | 'payment' | 'adjustment'
  amount: number
  note?: string
}

export type ActivityRow = {
  id: string
  date: string
  label: string
}

/** Aggregates from GET /customers/:id when provided by the API. */
export type CustomerLedgerStats = {
  totalPurchase: number
  amountPaid: number
  currentDebt: number
}

/** View model: API customer plus optional local ledger fields (empty until sales/credit APIs exist). */
export type CustomerRecord = {
  id: number
  name: string
  phone: string
  address: string
  note: string
  creditLimit: number | null
  status: CustomerStatus
  purchases: PurchaseRow[]
  debtLedger: DebtRow[]
  activities: ActivityRow[]
  createdAt: string
  updatedAt: string
  /** Present when loaded from customer detail endpoint (server-backed totals). */
  ledgerStats?: CustomerLedgerStats
}

export type CustomerMetrics = {
  totalPurchases: number
  currentDebt: number
  amountPaid: number
  lastPurchaseDate: string | null
  lastActivityDate: string | null
}

/** Matches backend `CustomerResponse` JSON (FastAPI / Pydantic). */
export type CustomerResponse = {
  id: number
  name: string
  phone: string
  address: string
  note: string
  credit_limit: string | number | null
  status: CustomerStatus
  created_at: string
  updated_at: string
}

/** Matches backend `CustomerPurchaseHistoryItem`. */
export type CustomerPurchaseHistoryItem = {
  sale_id: number
  total: string | number
  subtotal: string | number
  discount: string | number
  payment_method: string
  items_count: number
  created_at: string
}

/** Matches backend `CustomerDetailResponse` (GET /customers/{id}). */
export type CustomerDetailResponse = CustomerResponse & {
  debt_balance: string | number
  total_purchase: string | number
  amount_paid: string | number
  recent_purchases: CustomerPurchaseHistoryItem[]
}

export type CustomerCreatePayload = {
  name: string
  phone: string
  address: string
  note: string
  credit_limit: number | null
  status: CustomerStatus
}

export type CustomerUpdatePayload = CustomerCreatePayload

export function creditLimitFromApi(value: string | number | null | undefined): number | null {
  if (value == null || value === '') return null
  if (typeof value === 'number') return Number.isFinite(value) ? round2(value) : null
  const n = parseFloat(String(value))
  return Number.isFinite(n) ? round2(n) : null
}

function round2(n: number) {
  return Math.round(n * 100) / 100
}

File: backend/src/schemas/customer.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

CustomerStatus = Literal["active", "inactive", "watch"]


class CustomerBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    phone: str = Field(default="", max_length=50)
    address: str = Field(default="", max_length=2000)
    note: str = Field(default="", max_length=4000)
    credit_limit: Decimal | None = Field(default=None, ge=0, decimal_places=2)
    status: CustomerStatus = "active"


class CustomerCreate(CustomerBase):
    pass


class CustomerUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    phone: str | None = Field(default=None, max_length=50)
    address: str | None = Field(default=None, max_length=2000)
    note: str | None = Field(default=None, max_length=4000)
    credit_limit: Decimal | None = Field(default=None, ge=0, decimal_places=2)
    status: CustomerStatus | None = None


class CustomerResponse(CustomerBase):
    id: int
    created_at: datetime
    updated_at: datetime


class CustomerPurchaseHistoryItem(BaseModel):
    """A compact sale row for customer purchase history."""

    sale_id: int
    total: Decimal = Field(decimal_places=2)
    subtotal: Decimal = Field(decimal_places=2)
    discount: Decimal = Field(decimal_places=2)
    payment_method: str
    items_count: int
    created_at: datetime


class CustomerDetailResponse(CustomerBase):
    """Single-customer payload including purchase stats and recent sales."""

    id: int
    created_at: datetime
    updated_at: datetime
    debt_balance: Decimal = Field(decimal_places=2)
    total_purchase: Decimal = Field(
        decimal_places=2,
        description="Sum of sale totals for this customer (all payment methods).",
    )
    amount_paid: Decimal = Field(
        decimal_places=2,
        description="Sum of sale totals paid at checkout (cash, moncash, transfer only; credit sales excluded).",
    )
    recent_purchases: list[CustomerPurchaseHistoryItem]
